src/Prototype/Runtime/Errors.hs

```diff
@@ -0,0 +1,47 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE GADTs #-}
+module Prototype.Runtime.Errors
+  ( RuntimeErr(..)
+  , IsRuntimeErr(..)
+  ) where
+
+import qualified Network.HTTP.Types            as HTTP
+
+-- | A generalised error
+data RuntimeErr where
+  -- | Capture known error types. 
+  KnownErr ::IsRuntimeErr e => e -> RuntimeErr
+  -- | Capture all exceptions 
+  RuntimeException ::Exception e => e -> RuntimeErr
+
+-- | TODO: add common properties of errors. 
+class IsRuntimeErr e where
+
+  -- | Construct a `RuntimeErr` from an instance value  
+  knownErr :: e -> RuntimeErr
+  knownErr = KnownErr
+
+  -- | What should the error status be in HTTP-terms?
+  httpStatus :: e -> HTTP.Status
+
+  -- | Throw an error created with an instance value
+  throwError' :: MonadError RuntimeErr m => e -> m a
+  throwError' = throwError . knownErr
+
+  -- | Informative user message
+  -- TODO: also add langauge code as a parameter for localisation (later) 
+  userMessage :: e -> Maybe Text
+
+instance IsRuntimeErr RuntimeErr where
```

Comment:
I'm really having a hard time grasping this. So you have `RuntimeErr` that can be either "known" or not. But then the class `IsRuntimeErr` seems to be used to construct "known" error, and this implementation seems to say that all `RuntimeErr` can be wrapped by `knownErr`, and actually be 500...

Reply:
The idea is, say, for an error type X; we'd like to make an instance of `IsRuntimeErr`; we'd need to implement its instance. 

Consider: 

``` haskell 
data AuthErr = LoginFailed Text | PermissionDenied Text
             deriving Show
```

Which are two kinds of authentication failures; each having its specific error code. 

Thus, its instance will be: 

``` haskell 

instance IsRuntimeErr AuthErr where

  httpStatus = \case
    LoginFailed{} -> HTTP.unauthorized401
    PermissionDenied{} -> HTTP.forbidden403

  userMessage = Just . \case
    LoginFailed msg -> msg
    PermissionDenied msg -> msg

 ```

Now; in some `m` such that; `MonadError RuntimeErr m` we can then do: 

``` haskell 

do 
  mAuthd <- loginUser $ Credentials user pass 
  when (isNothing mAuthd) $ throwError' . LoginFailed $ show user <> " failed login." 
  -- etc. 

```

`throwError'` will wrap our error into a `KnownErr e :: RuntimeErr`; and at the servant mapping to servant errors; we'll map this out using the `IsRuntimeErr` instance of `RuntimeErr`.

I'll provide a more concrete example of this today.

Reply:
@noteed Please see: 
- https://github.com/noteed/start-servant/commit/0c7e185a960341d5eff9c0b920e399f597b6a194
- https://github.com/noteed/start-servant/commit/23212c715de578d45d49e48e4ca256543cf4a6c8

For the intended use case.